streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -272,6 +274,36 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
             }
         }
 
+        if (taskCloseExceptions.isEmpty()) {
+            Task oldTask = null;
+            final Iterator<Task> transitioningTasksIter = tasksToRecycle.iterator();
+            try {
+                while (transitioningTasksIter.hasNext()) {
+                    oldTask = transitioningTasksIter.next();
+                    final Task newTask;
+                    if (oldTask.isActive()) {
+                        final Set<TopicPartition> partitions = standbyTasksToCreate.remove(oldTask.id());
+                        newTask = standbyTaskCreator.createStandbyTaskFromActive((StreamTask) oldTask, partitions);
+                    } else {
+                        final Set<TopicPartition> partitions = activeTasksToCreate.remove(oldTask.id());
+                        newTask = activeTaskCreator.createActiveTaskFromStandby((StandbyTask) oldTask, partitions, mainConsumer);
+                    }
+                    tasks.remove(oldTask.id());
+                    addNewTask(newTask);
+                    transitioningTasksIter.remove();
+                }
+            } catch (final RuntimeException e) {
+                final String uncleanMessage = String.format("Failed to recycle task %s cleanly. Attempting to close remaining tasks before re-throwing:", oldTask.id());
+                log.error(uncleanMessage, e);
+                taskCloseExceptions.put(oldTask.id(), e);
+
+                dirtyTasks.addAll(tasksToRecycle); // contains the tasks we have not yet tried to transition
+                dirtyTasks.addAll(tasks.values());         // contains the new tasks we just created
```

Comment:
This will add all current tasks, is that right?

Reply:
Hm. The current logic is to only close dirty those tasks which were in a failed action (eg close, or in the case of a failed commit we will close all currently assigned active tasks), so I guess we never end up closing standbys, which makes sense. 

So maybe we shouldn't close any successfully recycled tasks at all, and just let them be closed during `handleLostAll` if necessary?